spectrum-offchain/src/box_resolver/persistence.rs

```diff
@@ -1,17 +1,404 @@
-    async fn get_prediction(&self, id: TEntity::TStateId) -> Option<Traced<Predicted<TEntity>>>;
-    async fn get_last_predicted(&self, id: TEntity::TEntityId) -> Option<Predicted<TEntity>>;
-    async fn get_last_confirmed(&self, id: TEntity::TEntityId) -> Option<Confirmed<TEntity>>;
-    async fn get_last_unconfirmed(&self, id: TEntity::TEntityId) -> Option<Unconfirmed<TEntity>>;
-    async fn put_predicted(&mut self, entity: Traced<Predicted<TEntity>>);
-    async fn put_confirmed(&mut self, entity: Confirmed<TEntity>);
-    async fn put_unconfirmed(&mut self, entity: Unconfirmed<TEntity>);
-    async fn invalidate(&mut self, eid: TEntity::TEntityId, sid: TEntity::TStateId);
+    async fn get_prediction<'a>(&self, id: TEntity::TStateId) -> Option<Traced<Predicted<TEntity>>>
+    where
+        <TEntity as OnChainEntity>::TStateId: 'a;
+    async fn get_last_predicted<'a>(&self, id: TEntity::TEntityId) -> Option<Predicted<TEntity>>
+    where
+        <TEntity as OnChainEntity>::TEntityId: 'a;
+    async fn get_last_confirmed<'a>(&self, id: TEntity::TEntityId) -> Option<Confirmed<TEntity>>
+    where
+        <TEntity as OnChainEntity>::TEntityId: 'a;
+    async fn get_last_unconfirmed<'a>(&self, id: TEntity::TEntityId) -> Option<Unconfirmed<TEntity>>
+    where
+        <TEntity as OnChainEntity>::TEntityId: 'a;
+    async fn put_predicted<'a>(&mut self, entity: Traced<Predicted<TEntity>>)
+    where
+        Traced<Predicted<TEntity>>: 'a;
+    async fn put_confirmed<'a>(&mut self, entity: Confirmed<TEntity>)
+    where
+        Traced<Predicted<TEntity>>: 'a;
+    async fn put_unconfirmed<'a>(&mut self, entity: Unconfirmed<TEntity>)
+    where
+        Traced<Predicted<TEntity>>: 'a;
+    async fn invalidate<'a>(&mut self, eid: TEntity::TEntityId, sid: TEntity::TStateId)
+    where
+        <TEntity as OnChainEntity>::TStateId: 'a,
+        <TEntity as OnChainEntity>::TEntityId: 'a;
+}
+
+static PREDICTED_KEY_PREFIX: &str = "predicted:prevState:";
+static LAST_PREDICTED_KEY_PREFIX: &str = "predicted:last:";
+static LAST_CONFIRMED_KEY_PREFIX: &str = "confirmed:last:";
+static LAST_UNCONFIRMED_KEY_PREFIX: &str = "unconfirmed:last:";
+
+#[async_trait(?Send)]
+impl<TEntity> EntityRepo<TEntity> for RedisClient
+where
+    TEntity: OnChainEntity + Clone + Serialize + DeserializeOwned,
+    <TEntity as OnChainEntity>::TStateId: Clone + Serialize + DeserializeOwned,
+    <TEntity as OnChainEntity>::TEntityId: Clone + Serialize + DeserializeOwned,
+{
+    async fn get_prediction<'a>(
+        &self,
+        id: <TEntity as OnChainEntity>::TStateId,
+    ) -> Option<Traced<Predicted<TEntity>>>
+    where
+        <TEntity as OnChainEntity>::TStateId: 'a,
+    {
+        let mut conn = self.pool.get().await.unwrap();
+        if let Ok(entity_bytes) = cmd("GET")
+            .arg(predicted_key_bytes(&id))
+            .query_async::<_, Vec<u8>>(&mut conn)
+            .await
+        {
+            match bincode::deserialize(&entity_bytes) {
```

Comment:
If the key doesn't exist in redis it will return (nil). Note that the error will be generated from `bincode::deserialize`, but we just ignore it and return `None`.